### ISSUE TITLE
do not expose the entire system (filesystem, devices)

### ISSUE THREAD

**Issue by anarcat** [flathub/com.calibre_ebook.calibre#51]
The permissions granted by this flatpak are really broad:

https://github.com/flathub/com.calibre_ebook.calibre/blob/2226c559fa169c26cb3ce68c14e94aff904ccc8e/com.calibre_ebook.calibre.yaml#L7-L15

`--filesystem=host` grants access to the entire filesystem. `--device=all` grants access to *any* device whatsoever, including virtualization (e.g. `kvm`)... isn't there a better knob here?

Alternatively, it would be nice if there were instructions on how to restrict this a little, for those of us not exactly confident in the security of calibre...

**Comment by ghisvail** [comment]
I am open to suggestions on this matter. Feel free to experiment and
provide feedback on what works and what does not.

**Comment by Erick555** [comment]
You may start by reading why those permissions were added in first place https://github.com/flathub/com.calibre_ebook.calibre/issues/13 https://github.com/flathub/com.calibre_ebook.calibre/pull/5 .

Flatpak permissions aren't fine grained so a lot of unnecessary stuff is allowed. You may remove them with `flatpak --user override --nodevice=all com.calibre_ebook.calibre` and `flatpak --user override --nofilesystem=host com.calibre_ebook.calibre`. Those are generic flatpak instructions and there are hundreds of flatpaks facing the same issue so I don't think each one should repeat the same info.

**Comment by OdinVex** [comment]
I specifically go out of my way to _remove_ all Flatpak restrictions when possible. The only, only reason I use Flatpak is for newer versions of software on distros that sadly hold back on software/libs/drivers/anything. I'd use an edge distro but I can't find a pretty Cinnamon distro. Anyway... I use **Flatseal** to control Flatpak stuff, such as adding (in your case, removing) permissions, etc.

**Comment by ghisvail** [comment]
I am happy to discuss a new set of Flatpak permissions. Feel free to provide suggestions.

**Comment by OdinVex** [comment]
I was merely commenting to suggest they could change the permissions as they wish, using Flatseal.

**Comment by Svadkos** [comment]
`device=dri` is enough. Also, calibre works with wayland now.
As for filesystem permissions - I don't know, `home` would be enough, but that way you won't be able to add books from external devices. Personally, I'll change filesystem permissions for my system to more restrictive in any case.

**Comment by ghisvail** [comment]
@Svadkos so I suppose you've tried with just `--filesystem=home` and `--device=dri` and verified it works with your ebook device of choice?

>  Also, calibre works with wayland now.

Meaning, changing to `--socket=wayland`. Again, did you check it? If so, could you submit a PR for it?

Cheers.

**Comment by Erick555** [comment]
`host` could be replaced with `home` `/run/media` `/media`. The difference won't be be significant from security pov though.

**Comment by anarcat** [comment]
> `host` could be replaced with `home` `/run/media` `/media`. The difference won't be be significant from security pov though.

on the contrary - `host` is huge and there's no reason to use it for an app like calibre. i would even say that `home` is too broad as it exposes executable code like `.bashrc` which makes remote code executions through a compromised calibre trivial.

**Comment by Erick555** [comment]
I meant `host` doesn't add much on top of `home` which gives more than enough to break any sandbox security guarantee. The problem with `home` is that without it calibre will be unable to find existing user library if it isn't in standard place like somewhere in xdg-documents. Removing it will break the app for many users overnight. I'm afraid that for security you need to tweak those perms by yourself (I do).

**Comment by ntninja** [comment]
Some notes from my side on this:

#### Wayland

Wayland works just fine and doesn't have problems with fractional DPI-scaling, please enable it!

(That is, replace `--socket=x11` by `--socket=wayland --socket=fallback-x11`.)

#### Filesystem access

Filesystem access isn't needed for just the eBook library (there may be more):

  1. Exporting eBooks opens a directory chooser and will attempt to place the generated eBook file(s) into a *subdirectory* of the selected directory → Only works in whitelisted directories
  2. Scanning mass-storage connected eBook device requires access to `/media`
  3. “Add books from directory and subdirectories”-option (could be acceptable to disable if that were the only issue here)

No access is required to “Add books from single directory” (uses multi-file select) or opening book files externally.

So, *if* forcing the eBook libray to a single directory is deemed acceptable *and* we can get a patch into Calibre that allows selecting the destination file (rather than directory) for saving exported files, then properly restricting this Flatpak *could* work at the expanse of breaking people's stuff *once*. Otherwise I don't see how this could work sadly.

#### `--talk-name=org.freedesktop.Notifications` D-Bus access

This one is really minor, but I fixed it nontheless (will be part of 5.5): https://github.com/kovidgoyal/calibre/pull/1268

#### `--system-talk-name=org.freedesktop.NetworkManager`

Less minor, but just as easy to fix as the notification access since it's only used for detecting online/offline mode and there exists an equivalent portal for this. – Will likely be fixed by me.

**Comment by OdinVex** [comment]
Grateful for the Wayland support with fallback for X11 but I disagree entirely for filesystem access unless the aim of Linux is to become a locked down stringy mess version of Mac/Windows. I don't see why any software (they're not called apps...) should be have restrictions placed on them out-of-box. Authority is restricted top-down, not ground-up.

**Comment by ntninja** [comment]
@OdinVex: Because „secure by default“ is a thing? Because it's better to start with small permissions and give users the ability to add more as they see fit, than leaving users to figure out after the fact which of the restriction buttons will and will not work?

Or do expect that a first-time Callibre user would know that restricting filesystem access will break the *export* feature, but will work just fine with *one of the two import* features!?
As I mentioned above, with the way Calibre works today, restricting filesystem access by default is unacceptable anyways. But when/if the situation improves it will make sense to start with a better default sandbox *for new users*. (Noone here, is arguing for pretty you from then extending the sandbox however you see fit!)

**Comment by OdinVex** [comment]
@ntninja A computer turned off and disassembled is secure by default. Unusable but whatever. New Users are on limited accounts, from top-down anyway. They're already sandboxed. No need to sandbox root because babies can't chew steaks. Root has already pre-cut it...

**Comment by Erick555** [comment]
> Because it's better to start with small permissions and give users the ability to add more as they see fit, than leaving users to figure out after the fact which of the restriction buttons will and will not work?

The problem is this flatpak didn't start with _small permissions_ and the _home_ access was there from the beginning. I think restricting filesystem access would be acceptable if it was done at time this flatpak was created when users could adjust the sandbox before they start using app. Making breaking changes for users during app update will seriously damage their trust for maintainers.

> Device/UDisks access
> 
> Needed, along with access to /media for accessing connected devices. Would need portals for physical device access and per-device filesystem access. Not fixable at this point.

I removed device access in app and I still can sent books to my kindle are there some more advanced functions that needs this permission?

> New Users are on limited accounts, from top-down anyway. They're already sandboxed. No need to sandbox root because babies can't chew steaks. Root has already pre-cut it...

Nothing is sandboxed in desktop linux without flatpak/snap/firejail, etc except maybe chromium which does self sandbox. Also flatpak is about sandboxing apps not users.

**Comment by OdinVex** [comment]
@Erick555 It was meant as a general statement. I never once thought Flatpak was about sandboxing. I have been told (until now) and always believed it to be a flat repository of a software to provide the latest updates and libraries regardless of the old-fart maintenance of distros.

Regardless, it won't bother me for long any way. I've tried for so long to undo a lot of the library/update crap of mainstream distros that I'm slowly switching to arch-based distros. Mainstream distros are popular mostly due to the same reasons Mac and Windows are popular but hanging onto old libs just doesn't do it for me anymore. Constantly providing work-arounds is taking up too much time to use new things I need, so Flatpak (never touched Snap, ugh) are out for me. Just thought I'd mention my piece on filesystem access to provide some kind of "other usable argument".

**Comment by ghisvail** [comment]
> But when/if the situation improves it will make sense to start with a better default sandbox for new users

I agree with this statement.

> to provide the latest updates and libraries regardless of the old-fart maintenance of distros

That was my motivation to start this Flatpak initially

> The problem is this flatpak didn't start with small permissions and the home access was there from the beginning

My mistake. Now can we move forward?

> Making breaking changes for users during app update will seriously damage their trust for maintainers.

If it provides better defaults for new and current users in the long run, I believe it's worth a shot.

Few people use Flatpak, even fewer use Flatseal.

I am glad people care enough about this Flatpak to have this conversation.

**Comment by ntninja** [comment]
> I removed device access in app and I still can sent books to my kindle are there some more advanced functions that needs this permission?

As far as I know, it can communicate with devices using either [MTP](https://en.wikipedia.org/wiki/Media_Transfer_Protocol) (uses it's own version of libmtp bundled with the app and direct USB access) or mass storage – it might do more than that, but I haven't investigated. Doing USB stuff from userland requires access to lots of stuff in `/sys` to enumerate devices and read device descriptors, but I'm fairly certain that you cannot actually *open* any raw USB devices from there but have to use `/dev` for that final part.

If your Kindle pops up as a disk device, then Calibre will just communicate with it by reading and writing files to that filesystem and will support it that way. As I said, it might have the ability to do *more* than that and that *more* may be blocked, but I cannot tell you anything about it. My only eReader device here only supports mass storage / disk device access anyways (PocketBook), so I cannot do any further tests on this either.

**Comment by ghisvail** [comment]
> My only eReader device here only supports mass storage / disk device access anyways

Same here (Bookeen Diva)

**Comment by OdinVex** [comment]
@ghisvail I have greatly enjoyed the Calibre Flatpak as a work-around to archaic libraries on Mint and although I have moved on to arch-based distros with AUR (holy grail in my opinion), I still care enough to respond regarding it, yeah. I found I needed network drives/filesystem/ext drives when I used Calibre. The first time I loaded it up in Flatpak trying to get updates that fixed issues, I ran into the filesystem permissions thing and flipped. It took two _months_ to figure out that Flatpak was sandboxing it. I immediately realized Flatpak was done for me and was no longer the solution it promised to be when first created. It had promised to just provide 'new libs' without DLL-hell. Nothing about Sandboxing. I imagine _that_ was added after the initial days but eh...it ruined Calibre for me almost. I was _lucky_ to find Flatseal to undo the damage of sandboxing as much as possible.

**Comment by Erick555** [comment]
> My mistake. Now can we move forward?

I didn't mean this was mistake, this was a choice that may be even preferred among flathub reviewers.

> If it provides better defaults for new and current users in the long run, I believe it's worth a shot.

The opinions vary here, given how much home/host permission are popular on flathub I think the majority is for what calibre does currently.

> Few people use Flatpak, even fewer use Flatseal.

Thousands of users use this flatpak right now. I'm afraid significant amount of them will be upset if working app stops working overnight and need some action to unbreak it.

```
Counting updates between 2020-10-16 07:53:24 +0000 and 2020-10-30 15:14:19+01:00:
............... done!

x86-64 (98%): 17766 downloads including 14760 updates
i386 (1%): 286 downloads including 284 updates
aarch64 (0%): 0 downloads including 0 updates
Arm (0%): 0 downloads including 0 updates

Total:   18052 downloads in 15 days
Average: 1203.46 downloads per day

```

**Comment by ntninja** [comment]
As I've come to discover, (a) [the portal documentation at docs.flatpak.org](https://docs.flatpak.org/en/latest/portal-api-reference.html#gdbus-method-org-freedesktop-portal-FileChooser.OpenFile) is pretty out of date, (b) since version 3 of the FileChooser portal an app [can ask the runtime to open a directory of a file](https://github.com/flatpak/xdg-desktop-portal/commit/c616b491fe0ac67f78a14bcaf0d3a150bfadfd82) and was included in XDP 1.8, (c) since version 2 of the FileChooser portal an app can ask the runtime to save several files at once.

It appears that with these facilities the ideal path for migration becomes possible without breaking people's existing libraries needlessly:

> Ideally, that directory portal work would finally come along. Then Calibre could simply ask where it should place it's eBook library on first start and implicitly gain access to the corresponding directory by that choice. It would also make transition for existing users painless as we could simply tell them about this on first start after the transition and ask them to confirm their existing eBook location by browsing to that location once.

There are some issues to be considered before this can work however:

  1. Is the Qt version used compatible with version 3 of the file portal? (Without this it will use its internal file chooser and things will be broken.)
  2. Are the versions of `xdg-desktop-portal{,-gtk,-kde}` on the target systems recent enough to expose this feature? (Ubuntu 20.04LTS was released with `xdg-desktop-portal` version 1.6 for instance.)
  3. Oh, and somebody will have to write to relevant patches for Calibre of course…

@Erick555: Are there also numbers available which OS users of this Flatpak are using?

**Comment by Erick555** [comment]
Yes newer portal can open directories but the problem is that it needs explicit user action - when you remove host permission then calibre will lose access to library and users will have to re-add it manually (assuming the whole functionality is working).

> Are there also numbers available which OS users of this Flatpak are using?

No, flatpak doesn't leak such info.

**Comment by ntninja** [comment]
> Yes newer portal can open directories but the problem is that it needs explicit user action - when you remove host permission then calibre will lose access to library and users will have to re-add it manually (assuming the whole functionality is working).

If it means users will be asked once after upgrade to “confirm” their existing library location, by navigating to it using the file chooser I consider that an acceptable trade-off. The permission granted by this will be persisted mind you.

So I wanted to test this theory, by removing all Calibre filesystem permissions and manually granting it access to the eBooks folder (I have all the recent versions needed for this). Unfortunately it didn't work out quite the way I imagined:

  1. On the bright side: It appears that the Qt version in my Calibre from-source build actually has support for the new directory chooser facility, so that part just worked and the directory made available with no issues or code changes, just by selecting it from the first-run wizard.
  2. Also, the directory permission was persisted across restarts at the same path, so no special code needed there either.
  3. HOWEVER, the library also contains a SQLite3 database with Calibre's index which fails to initialize from the shared directory path. Causing Calibre to first report database corruption, then crashing after rebuilding the database.

Summary: Look's promising, but without some way to cause it to *not* crash on that database this will not work. (A simple idea would be to move the database into `${XDG_DATA_HOME}/calibre` instead, but maybe we can do better.)

**Comment by ntninja** [comment]
> No, flatpak doesn't leak such info.

Sad (for our purposes here), but not unexpected or even bad in general.

**Comment by Erick555** [comment]
> If it means users will be asked once after upgrade to “confirm” their existing library location, by navigating to it using the file chooser I consider that an acceptable trade-off. The permission granted by this will be persisted mind you.

The users won't know what happened and what they are expected to do as there is now way to communicate this (we can't assume they follow changes on github).

**Comment by ntninja** [comment]
> there is now way to communicate this (we can't assume they follow changes on github).

Wrong, we can just throw up a message dialog telling them what happened, then they don't need to be confused…

**Comment by ghisvail** [comment]
> there is now way to communicate this

Technically, we could with a wrapper script launching a zenity message box. That's what is done for games requiring separate install of proprietary assets which cannot be bundled or installed as extra-data within the Flatpak.

**Comment by ghisvail** [comment]
Has anyone tried launching the app in a wayland session? It does not work for me:

```
$ flatpak run com.calibre_ebook.calibre
 
Warning: Ignoring XDG_SESSION_TYPE=wayland on Gnome. Use QT_QPA_PLATFORM=wayland to run on Wayland anyway.
qt.qpa.xcb: could not connect to display 
qt.qpa.plugin: Could not load the Qt platform plugin "xcb" in "" even though it was found.
This application failed to start because no Qt platform plugin could be initialized. Reinstalling the application may fix this problem.

Available platform plugins are: eglfs, linuxfb, minimal, minimalegl, offscreen, vnc, wayland-egl, wayland, wayland-xcomposite-egl, wayland-xcomposite-glx, xcb.
```

That's on Debian 10.

**Comment by Erick555** [comment]
> Wrong, we can just throw up a message dialog telling them what happened, then they don't need to be confused…

I consider patching app or adding special wrapper for this as quite invasive change.

**Comment by ntninja** [comment]
> Has anyone tried launching the app in a wayland session? It does not work for me:

Works just fine on KDE/Wayland for me. There seems to be some weird GNOME-only only rule in QtWayland for this[1](https://discussion.fedoraproject.org/t/ignoring-xdg-session-type-wayland-on-gnome/6082). My from-source build does not appear to be affected by this, but that might be KDE patch. So for now, just try doing what it says and start it with `QT_QPA_PLATFORM=wayland` to test it.

> Technically, we could with a wrapper script launching a zenity message box.

I don't consider a wrapper script to be too invasive, but without @Erick555's input I probably would have been too lazy to actually do such a thing. :+1:

> Will it show up for every launch then or how do you detect if user already fixed the problem?

After taking a quick look the migration logic can be as simple as this:

  1. Open `${XDG_CONFIG_HOME}/calibre/global.py.json` and parse it as JSON
  2. Check the value of the `"library_path"` key:
       1. If it starts with `${XDG_RUNTIME_DIR}/`, go to step 5 (path is already managed by document portal)
       2. It it refers to an existing directory, go to step 5
  3. Open dialog explaining the situation and ask users to use the on-screen file chooser button to open their previous eBook library path that is displayed on the screen
  4. Change value of `"library_path"` in the JSON and write back the updated `global.py.json` file
  5. Start Calibre as usual

**Comment by Erick555** [comment]
> My from-source build does not appear to be affected by this, but that might be KDE patch.

yes, kde flatpak runtime [changes this](https://invent.kde.org/packaging/flatpak-kde-runtime/-/blob/qt5.15lts/patch/qtbase-use-wayland-on-gnome.patch). You may consider changing `fallback-x11` to` x11`  to fix non-source build.

**Comment by sat3ll** [comment]
> 
> > Device/UDisks access
> > Needed, along with access to /media for accessing connected devices. Would need portals for physical device access and per-device filesystem access. Not fixable at this point.
> 
> I removed device access in app and I still can sent books to my kindle are there some more advanced functions that needs this permission?
> 

I've added that permission since my Kindle Fire (MTP device) would refuse to work without it.

**Comment by anarcat** [comment]
i ended up overriding things in my executable as well. i have this override:

```
[Context]
devices=!all;
filesystems=!host;~/.config/calibre/;/srv/books-incoming/;~/books-incoming;/srv/books/;~/books/;xdg-documents;xdg-download;xdg-desktop;
```

but somehow it doesn't work: when I try to add a book to my library i can browse the filesystem (and, surprisingly, all of it) and add the file, but then progress dialog doesn't pop up and the file doesn't get added to the library.

i suspect this might have been because of the incorrect `database_path` in `global.py.json`, but it doesn't seem to be it.

looking at those config files a little further, it's kind of a nightmare anyways: the `global.py.json` seems to be a copy of the `global.py` config file, which is just a Python script. hello code execution... 

seems kind of impossible / pointless to secure this thing at this point...

**Comment by Erick555** [comment]
From which directory you tried to add this file? Also what distro did you use and what xdg-dsktop-portal version it has?

The file chooser works through portal which is opened outside sandbox thus it can see all your system.

`~/.config/calibre` doesn't make sense as calibre flatpak config resides under `~/.var`.

I can only tell you that I use calibre flatpak without host permission and it works very well for me so securing it is definitely doable.

**Comment by anarcat** [comment]
> From which directory you tried to add this file?

The override is placed in `/var/lib/flatpak/overrides/com.calibre_ebook.calibre`. I'm certain it works, because if i remove it, calibre works again. :)

>  Also what distro did you use and what xdg-dsktop-portal version it has?

Debian stable, xdg-desktop-portal 1.2.0-1.

> ~/.config/calibre doesn't make sense as calibre flatpak config resides under ~/.var.

Hum. maybe that's the problem here... The thing is my calibre config *is* inside `~/.config/calibre`... or did it move?

> I can only tell you that I use calibre flatpak without host permission and it works very well for me so securing it is definitely doable.

Which config are you using?

**Comment by Erick555** [comment]
> The override is placed in /var/lib/flatpak/overrides/com.calibre_ebook.calibre. I'm certain it works, because if i remove it, calibre works again. :)

I meant from which dir you tried to add book to library.

> Debian stable, xdg-desktop-portal 1.2.0-1.

That's quite old, I'm not sure how reliably it works. It's possible that it mises some fixes but I don't know.

> Hum. maybe that's the problem here... The thing is my calibre config is inside ~/.config/calibre... or did it move?

All flatpaks set `XDG_CONFIG_HOME=~/.var/app/<appid>/config` so flatpak calibre (which follows xdg spec) should use `~/.var/app/com.calibre_ebook.calibre/config` to store its config unless it get somewhat hardcoded `~/.config/calibre` path. Are you sure your calibre flatpak uses `~/.config/calibre` and not `~/.var/app/com.calibre_ebook.calibre/config/calibre`? Does the latter exist?

> Which config are you using?

My calibre flatpak has its config in `~/.var/app/com.calibre_ebook.calibre/config`. It also has access only to books dir under my home, xdg-downloads, and /run/media (for removable storage access).

**Comment by ntninja** [comment]
> but somehow it doesn't work: when I try to add a book to my library i can browse the filesystem (and, surprisingly, all of it) and add the file, but then progress dialog doesn't pop up and the file doesn't get added to the library.

Does it output anything useful if you start it using `flatpak run com.calibre_ebook.calibre` from terminal before doing that? (The `xdg-desktop-portal` version should be fine, but there could be bugs of course.)

You can get more debugging output by also starting XDP stuff yourself:

```sh
/usr/libexec/xdg-desktop-portal --replace -v &
/usr/libexec/xdg-document-portal --replace -v &

# For GTK portal (non-KDE) only
/usr/libexec/xdg-desktop-portal-gtk --replace -v &

# For KDE only
killall -w /usr/lib/x86_64-linux-gnu/libexec/xdg-desktop-portal-kde && /usr/lib/x86_64-linux-gnu/libexec/xdg-desktop-portal-kde &

flatpak run com.calibre_ebook.calibre
```
